### PR TITLE
fix(nextly): fall back for empty slugs and re-sanitize hook-set slugs

### PR DESCRIPTION
**Pull request by mobeenabdullah**
## What

Two follow-up fixes for slug auto-generation, surfaced by Codex's post-merge review of #222.

### 1. Empty slug when the title has no URL-safe characters
`generateSlug` strips everything outside `[\w-]`, so a title that is entirely non-ASCII, emoji, or punctuation (e.g. `create({ data: { title: "你好世界" } })`) sanitized to `""`. The previous fallback token only fired when the title/name was *missing*, not when derivation *yielded* empty, so the create failed required-field validation with an empty `slug`.

Now derivation falls back to a unique `entry-<ts>-<rand>` token whenever the sanitized base is empty. This also covers an **explicit** slug that sanitizes away (e.g. `slug: "你好"`), which is treated as unset and derived instead of persisting `""`.

### 2. Re-sanitize a slug set by a field `beforeValidate` hook
#222 moved slug generation ahead of field-level write access + validation. Field-level `beforeValidate` hooks run *after* that step, so a slug field hook could set an unsanitized value (`"Hello World"`) that was then validated and stored verbatim — a regression from the pre-#222 ordering, which sanitized after those hooks. All three create paths (`createEntry`, `createEntryInTransaction`, `createSingleEntryInTransaction`) now re-sanitize `slug` after `beforeValidate` hooks and before validation. The pass is idempotent for an already-clean slug and only overwrites when sanitization yields a non-empty value, so a hook that clears the slug still surfaces as the normal required-field error.

## Tests
Two integration tests added to `slug-autogen.integration.test.ts`:
- CJK-only title falls back to an `entry-` token.
- A slug field `beforeValidate` hook returning `"Hooked Slug Value"` is stored as `hooked-slug-value`.

Full slug suite (6) + plugin integration suite (37 files / 86 tests) green; `check-types` + `lint` clean.

## Not in this PR
A third Codex finding — the adapters' `TransactionContext` read methods (`selectOne`/`select`/…) delegate to the main connection rather than the transaction, so tx-scoped slug dedupe is only effective on SQLite — is an adapter-layer change tracked as a separate follow-up.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved slug auto-generation so slugs that sanitize to empty (including non-string results) are treated as unset and never stored as `""`; a unique fallback is derived instead.
  * Re-sanitizes/normalizes `slug` after both validation and change hooks during create flows, including transactional paths, to ensure consistent URL-safe storage.
* **Tests**
  * Expanded integration coverage for slug sanitization and hook edge cases across create scenarios.

### CODE REVIEW

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 1]
```diff
@@ -411,6 +413,45 @@ export class CollectionMutationService extends BaseService {
...
+   * validated and stored verbatim. Normalizing here keeps the stored slug
+   * URL-safe; it is idempotent for an already-clean slug and only overwrites
+   * when sanitization yields a non-empty value, so a hook that clears the slug
+   * still surfaces as the normal required-field validation error.
+   */
+  private reSanitizeSlug(finalData: Record<string, unknown>): void {
+    if (typeof finalData.slug === "string" && finalData.slug.trim() !== "") {
+      const sanitized = generateSlug(finalData.slug);
+      if (sanitized !== "") finalData.slug = sanitized;
+    }
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Good catch, fixed in 089d9318. You're right that leaving the raw value stored an un-sanitized slug verbatim. reSanitizeSlug now falls back to deriveSlug (derive from title, then the entry-<ts>-<rand> token) when the hook value sanitizes to empty, exactly like applyGeneratedSlugAndTitle does for an explicit slug that sanitizes away. Added a regression test: a beforeValidate hook returning "你好世界" now stores the title-derived slug, never the raw CJK value.

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 2]
```diff
@@ -411,6 +413,45 @@ export class CollectionMutationService extends BaseService {
...
+  }
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Good catch, fixed in 089d9318. You're right that leaving the raw value stored an un-sanitized slug verbatim. reSanitizeSlug now falls back to deriveSlug (derive from title, then the entry-<ts>-<rand> token) when the hook value sanitizes to empty, exactly like applyGeneratedSlugAndTitle does for an explicit slug that sanitizes away. Added a regression test: a beforeValidate hook returning "你好世界" now stores the title-derived slug, never the raw CJK value.

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 3]
```diff
@@ -411,6 +413,45 @@ export class CollectionMutationService extends BaseService {
...
+   * the title) could introduce an unsanitized value that would otherwise be
+   * validated and stored verbatim. Normalizing here keeps the stored slug
+   * URL-safe; it is idempotent for an already-clean slug and only overwrites
+   * when sanitization yields a non-empty value, so a hook that clears the slug
+   * still surfaces as the normal required-field validation error.
+   */
+  private reSanitizeSlug(finalData: Record<string, unknown>): void {
+    if (typeof finalData.slug === "string" && finalData.slug.trim() !== "") {
+      const sanitized = generateSlug(finalData.slug);
+      if (sanitized !== "") finalData.slug = sanitized;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Good catch, fixed in 089d9318. You're right that leaving the raw value stored an un-sanitized slug verbatim. reSanitizeSlug now falls back to deriveSlug (derive from title, then the entry-<ts>-<rand> token) when the hook value sanitizes to empty, exactly like applyGeneratedSlugAndTitle does for an explicit slug that sanitizes away. Added a regression test: a beforeValidate hook returning "你好世界" now stores the title-derived slug, never the raw CJK value.

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 4]
```diff
@@ -2824,6 +2873,10 @@ export class CollectionMutationService extends BaseService {
         });
       }
 
+      // A beforeValidate hook can set `slug` after generation ran; re-sanitize
+      // so the validated and stored value stays URL-safe.
+      this.reSanitizeSlug(finalData);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Fixed in 089d9318. Moved the reSanitizeSlug call inside the if (!skipHooks) block (both the beforeValidate and the new beforeChange one), so it only runs when a hook actually executed. When hooks are skipped the slug is still the already-sanitized generated value, so no pass is needed. Comment updated to match.

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 5]
```diff
@@ -594,6 +635,10 @@ export class CollectionMutationService extends BaseService {
         user: params.user,
       });
 
+      // A beforeValidate hook can set `slug` after generation ran; re-sanitize
+      // so the validated and stored value stays URL-safe.
+      this.reSanitizeSlug(finalData);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Confirmed and fixed in 089d9318. You're right that a create-time beforeChange hook ran after the only sanitize pass and could set an unsafe stored slug, unlike the update path which sanitizes after beforeChange. The three create paths now call reSanitizeSlug after the beforeChange hooks as well (with the same empty-sanitize to deriveSlug fallback). Added a regression test for a slug beforeChange hook returning "Changed By Hook" stored as "changed-by-hook".

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 6]
```diff
@@ -411,6 +413,53 @@ export class CollectionMutationService extends BaseService {
...
+  private async reSanitizeSlug(
+    finalData: Record<string, unknown>,
+    isSlugTaken: (slug: string) => Promise<boolean>
+  ): Promise<void> {
+    if (typeof finalData.slug === "string" && finalData.slug.trim() !== "") {
+      const sanitized = generateSlug(finalData.slug);
+      finalData.slug =
+        sanitized !== ""
+          ? sanitized
+          : await this.deriveSlug(finalData, isSlugTaken);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Fixed in 933f62a1. reSanitizeSlug no longer skips blank/whitespace slugs: any value that sanitizes to empty (blank or non-URL-safe alike) is now re-derived from the title, so a beforeChange hook that clears the slug can't leave an empty value stored after validation has passed. Added a regression test: a beforeChange hook returning "   " now stores the title-derived slug ("fallback-title").

File: packages/nextly/src/plugins/__tests__/slug-autogen.integration.test.ts [hunk 7]
```diff
@@ -90,4 +90,97 @@ describe("slug auto-generation on create", () => {
...
+    });
+    current = await createTestNextly({ collections: [posts] });
+
+    const created = await current.nextly.create({
+      collection: "posts",
+      data: { title: "Ignored" },
+    });
+
+    expect((created.item as { slug?: string }).slug).toBe("changed-by-hook");
+  });
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/plugins/__tests__/slug-autogen.integration.test.ts:185]: Fixed in 933f62a1. reSanitizeSlug no longer skips blank/whitespace slugs: any value that sanitizes to empty (blank or non-URL-safe alike) is now re-derived from the title, so a beforeChange hook that clears the slug can't leave an empty value stored after validation has passed. Added a regression test: a beforeChange hook returning "   " now stores the title-derived slug ("fallback-title").

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 8]
```diff
@@ -411,6 +413,56 @@ export class CollectionMutationService extends BaseService {
...
+   * sanitizes to empty (a CJK/emoji/punctuation-only string), it derives a
+   * valid slug from the title just like `applyGeneratedSlugAndTitle` does for
+   * an explicit slug that sanitizes away, rather than leaving the un-sanitized
+   * value to be stored verbatim.
+   */
+  private async reSanitizeSlug(
+    finalData: Record<string, unknown>,
+    isSlugTaken: (slug: string) => Promise<boolean>
+  ): Promise<void> {
+    if (typeof finalData.slug !== "string") return;
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts]: Fixed in c232f5fb. reSanitizeSlug no longer returns early for non-string slugs. Since applyGeneratedSlugAndTitle always sets a string slug before hooks run, any non-string value here means a hook set it; the guard now treats anything that isn't a non-empty sanitizable string (blank, non-URL-safe, or non-string like null) as needing a derived slug. Verified the full plugin integration suite (90 tests) still passes, so this doesn't force slugs onto collections that lack the field — applyGeneratedSlugAndTitle already establishes the string slug. Added a regression test for a beforeChange hook returning null.

File: packages/nextly/src/domains/collections/services/collection-mutation-service.ts [hunk 9]
```diff
@@ -594,6 +647,10 @@ export class CollectionMutationService extends BaseService {
         user: params.user,
       });
 
+      // A beforeValidate hook can set `slug` after generation ran; re-sanitize
+      // so the validated and stored value stays URL-safe.
+      await this.reSanitizeSlug(finalData, isSlugTaken);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts:660]: Good catch — this was a real bug I introduced, fixed in 87c1275c. reSanitizeSlug now returns early when `finalData.slug === undefined`. Field-level write access strips a denied field by DELETING the key, so undefined means "stripped by access (or never set)" and must be left alone — deriving would smuggle the slug back past access control. A slug the caller cannot create now stays stripped and the required-field validation applies, exactly as before slug re-sanitization was added. A PRESENT value (a hook set a string/blank/non-string) is still normalized or derived.

[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/nextly/src/domains/collections/services/collection-mutation-service.ts:660]: Addressed in 87c1275c, with a deliberate choice I want to flag. The access-stripped case (undefined) now short-circuits, so a slug removed before validation is no longer re-derived and the required check applies. For a hook that sets slug to `""`/whitespace, I'm keeping the derive rather than "letting validation fail", because I verified that `validateEntryData` currently permits an empty string for a required field (an all-emoji hook value sanitizes to `""` and is NOT rejected) — so "letting validation fail" would actually persist an empty slug, which is worse. Deriving yields a valid slug, mirroring applyGeneratedSlugAndTitle's empty fallback. The empty-string-passes-required gap in validateEntryData is pre-existing and broader than slug; I'll track it as a separate follow-up rather than special-case it here. Happy to switch to hard-rejecting an empty required slug if you'd prefer that globally.
